[spmd] Introduce Compile Mode FSDP with DTensor

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99374
* #99489
* #99436
* #99373
* #99435
* __->__ #99062
* #98452

This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

Differential Revision: [D45174400](https://our.internmc.facebook.com/intern/diff/D45174400)

wanchaol committed:
[spmd] Introduce Compile Mode FSDP with DTensor

This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+                    reshard_gm_nodes = list(reshard_gm.graph.nodes)
+                    input_node = reshard_gm_nodes[0]
+                    with graph.graph.inserting_before(node):
+                        output_node = graph.graph.graph_copy(
```
looks like `graph_copy` will return a list of arg nodes. Do we need to unpack that?

wanchaol replied:
hmmm I didn't see any issue without unpacking, probably because if the copying gm have a single output it would return a single output instead.

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+    exactly the sharding layout of intermediate tensors.
+
+    We possibly should use symbolic shapes to track the batch dimension, but this
+    needs to happen with dynamo, as dynamo is the only place we can mark a single
```
could you please elaborate on this one? We can get symints in on the traced graph of local train step, will that be sufficient if we never flatten batch dimension with other dimensions? In some RecSys models, we saw batch dim on the 3rd dimension in the overall input tensor. We can ask users to mark that and track it all the way through symints if necessary.

Curious, why Dynamo can faithfully identify which dim is batch dimension?

wanchaol replied:
that would be sufficient, what I am referring to here is that we probably don't want to mark input dimension as sym ints and might only want to mark the batch dimension, currently only dynamo gives this API `mark_dynamic` which allow us to only mark certain tensor dimensions. 

As discussed in the meeting, if symbolic shape in some case can't do cuda graph lowering and drags the performance, we might want to write our own batch dimension analysis in a more canonical way.

ezyang replied:
This claim "dynamo is the only place we can mark a single dimension" is simply untrue.

Here is an example script that does what you want:

```
import torch
from torch.fx.experimental.symbolic_shapes import ShapeEnv, DimDynamic
from torch._subclasses.fake_tensor import FakeTensorMode
        
shape_env = ShapeEnv()
fake_mode = FakeTensorMode(shape_env=shape_env)
        
t = torch.randn(4, 3)
fake_t = fake_mode.from_tensor(t, dynamic_dims=[DimDynamic.DYNAMIC, DimDynamic.STATIC])
print(fake_t.shape)

print(fake_t.sum(dim=0).shape)
print(fake_t.sum(dim=1).shape)
```

this prints

```
torch.Size([s0, 3])
torch.Size([3])
torch.Size([s0])
```

What I will say is that the API surface for direct FakeTensorMode manipulation isn't all that well developed. If you can collect some requirements for us there are probably some easy improvements for your use case.

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+        # do full reduction across batch dimension, it would still
+        # keep the reduction activation sharded.
+        reduction_over_batch = False
+        reduction_ops = [aten.sum.default, aten.mean.default]
```
this is great! will solve a lot of pain

mrshenli replied:
and we gonna need many more loss ops been marked as local reduction as well

wanchaol replied:
Yes! Do you know what loss ops we are using currently? I can add them to this list first

mrshenli reviewed (commented):
Thanks @wanchaol! This is awesome! I did a quick read and will get back later. 

Just to make sure we are aligned: would I be correct that DataParallelMode will buy us time to avoid chasing down a lot of op coverage issue when supporting DDP/FSDP, and will also help us to avoid the issue on local op/reduction. But we still need to harden DTensor-based expansion because we will need to add TP and sequence parallel later on and make them composable with DDP/FSDP?

kumpera reviewed (changes requested):
I think "symbolic" tracing is a bad idea as it forces us into using an even more unstable part of the PT2 stack. We're beyond our innovation budget already.

Beyond that, the usage of torch.library will not work with torch.deploy and needs to be fixed.

wconstab commented on torch/distributed/_spmd/api.py:
```diff
@@ -429,22 +445,59 @@ def stateless_func(func, params, buffers, named_states, args, kwargs):
+    # Otherwise it hits shape mismatch error, as we use local inputs to
+    # trace local graph and use DTensor to expand operators, where
+    # DTensor's shape is the global shape.
+    tracing_mode = "fake" if is_data_parallel_mode else "symbolic"
```
there is a distinction to draw here.

Using symbolic tracing mode helps us produce safe traces, that's all.  We start with real inputs, and if we don't call any strange ops like nonzero we won't get any symints along the way either.  we still get to do inductor optimizations relative to the parts of the trace that are real, which in your case with these typical models, is probably the whole model.  The upside of using symbolic tracing is, if some op in a model does produce a symint, we safely trace that instead of producing an unsafe trace and not knowing it.

Enabling dynamic shapes in PT2 is different.  Its first version started with all input dims being symints, and producing a hard-to-optimize program.  The most recent incarnation does a trick called auto dynamism, where it actually traces/compiles using static input shapes on the first try, then notices by observing which guards failed on iter2, and retraces with only those input dims as dynamic.

wanchaol replied:
Yeah I think it depends on whether the PT2 want to enable dynamic shape by default in the future as that would be a safe trace as @wconstab mentioned, we don't really need to buy in symbolic shape at this point. 

About what I am doing in this PR, I want to clarify: This PR does not use symbolic tracing, the previous DTensorExpand approach used symbolic tracing to deal with global shape/local shape mismatch issue. This PR just tries to preserve the old behavior and it uses "fake" mode tracing instead which produce static shapes.

wanchaol commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -50,6 +54,75 @@ def transform_and_compile(self, gm: GraphModule) -> GraphModule:
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> GraphModule:
+        # TODO: figure out a way to avoid explicit "cuda" mesh.
```
Yeah we would need to discuss more on this, taking a explicit mesh as argument is an way, but we might want to create a mesh automatically later if it's a complicated parallelism (i.e. 2-d parallel).

wanchaol commented on torch/distributed/_spmd/api.py:
```diff
@@ -429,22 +445,59 @@ def stateless_func(func, params, buffers, named_states, args, kwargs):
+            # For DataParallel mode, install hooks first to tag the graidents
+            # we would remove those nodes later, this is safe to trace
+            tagging_hooks = []
+            if isinstance(parallel_mode, DataParallel):
```
Sure I'll try put this into a context manager instead.

wanchaol commented on torch/distributed/_spmd/api.py:
```diff
@@ -24,19 +24,35 @@
 from torch.nn.utils._named_member_accessor import NamedMemberAccessor
 
 
+# used by data parallel to tag gradients.
```
Does this mean we can't have any custom op registration in python, even if it lives under spmd folder?

wanchaol commented:
> Just to make sure we are aligned: would I be correct that DataParallelMode will buy us time to avoid chasing down a lot of op coverage issue when supporting DDP/FSDP, and will also help us to avoid the issue on local op/reduction. 

@mrshenli Yes! This DataParallelMode would help us get rid of all those annoying issues I believe, and we don't even maintain any intermediate DTensor activations, which I think it would make compilation a lot faster.

> But we still need to harden DTensor-based expansion because we will need to add TP and sequence parallel later on and make them composable with DDP/FSDP?

Yeah as discussed in the meeting, we would need DTensor ops directly when it comes to TP or Sequence Parallel, so we might need to harden how to do DTensor Expansion, but I do think it would be a different implementation compare to the current one.

wanchaol commented:
> I think "symbolic" tracing is a bad idea as it forces us into using an even more unstable part of the PT2 stack. We're beyond our innovation budget already.

@kumpera already commented, but thought I should clarify again: This PR does not buy in "symbolic" tracing, the previous DTensorExpand approach used symbolic tracing to deal with global shape/local shape mismatch issue. This PR just tries to preserve the old behavior and it actually uses "fake" mode tracing instead which produce static shapes.

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

awgu commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+        return local_shard
+
+    for idx, placement in enumerate(spec.placements):
+        if placement.is_shard():
```
Curious: Can this only be `True` for one placement in `spec.placements`?

wanchaol replied:
not really, we can have something like `[Shard(0), Shard(1)]` on a 2-d mesh

awgu replied:
I misread the implementation. You continue sharding `local_shard` for each placement. This makes sense now :)

awgu commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+        assert self.batch_dim_size != -1, "batch dim size is not initialized!"
+
+        if node in self.batch_dim_map:
+            node_batch_dim = self.get_batch_dim(node)
```
Curious: What is the mental model when writing distributed compiler passes? Does CPU overhead matter? If compilation time is high, then we look into it?

wanchaol replied:
Yeah like you said, we should keep the overhead low when writing passes. I think mostly when the compilation takes long we investigate which pass and try to optimize it. My general feeling is that we should avoid writing expensive passes (at least not turn expensive ones on by default)

awgu commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+
+    def get_batch_dim(self, node: fx.Node) -> int:
+        if node not in self.batch_dim_map:
+            raise RuntimeError(f"batch dim analysis failed on node: {node}!")
```
Do we ever expect this to fail? Or, is this indicative of an internal implementation issue with `BatchDimAnalyzer`?

wanchaol replied:
This is most likely just an indicative of an internal implementation issue so that we can fix it.

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

awgu commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,731 @@
+    flattened_states = pytree.tree_flatten(named_states)[0]
+    num_states = len(flattened_states)
+    # 1. First build up data parallel strategies for the whole graph
+    strategy_map = build_data_parallel_strategies(
```
For now, we assume batch dimension is alway dim-0?

How do we get the batch dimension in general if it is not dim-0?

wanchaol replied:
Yeah usually when doing ddp/fsdp we have this implicit assumption that batch dim is 0. However for those cases that aren't batch dim = 0, in batch dim analyzer we have a field to record batch dim, it currently infer from the first input directly, but if user pass in a custom batch dim, we might can adjust that. It's really hard to know the batch dim if user don't provide a hint though, I think this can be a field of `DataParallelMode`

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+        # find a param_grad node via the tagging
+        if node.target == torch.ops._spmd.tag_grad.default:
+            cur_node = node
+            while cur_node.target in non_compute_ops:
```
`tag_grad` is in `non_compute_ops` list, but I assume we don't expect `tag_grad` has an ancestor `tag_grad` node?

wanchaol replied:
Yeah we don't expect that, the intention of putting `tag_grad` to non_compute_ops list is that we need to at least run this while loop once so that we can mark the input to `tag_grad` be partial (but python don't have `do...while` semantic).

ezyang replied:
Probably still better to factor the loop body into a function, and then do body(); while cond: body()

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+            cur_node = node
+            while cur_node.target in non_compute_ops:
+                cur_node = cur_node.args[0]
+                partial_strategy = _gen_partial_strategy(mesh)
```
nit: do we expect mutations on `partial_strategy`? If not, this can be shared across all grad nodes?

mrshenli replied:
I was curious about whether we can treat `partial_strategy` as a constant? If yes, we only need one of that instead of creating a new one for every node?

wanchaol replied:
I see, good point! Let me think about this, maybe `replicate` and `partial` strategy can both be shared, only shard spec need to be generated separately

ezyang replied:
If you are going to do this, I recommend turning the strategies into frozen dataclass, then you can singleton-ify them and the freeze will prevent people from trying to mutate the strategy.

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                break
+
+    train_step_graph.recompile()
+    train_step_graph.print_readable()
```
nit: debugging vestige

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+    for node in train_step_graph.graph.nodes:
+        if node.op == "placeholder":
+            if placeholder_idx < num_params:
+                shard_sig = _gen_shard_strategy(mesh, 0)
```
curious, what does the `sig` suffix mean?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+        if node.op == "placeholder":
+            if placeholder_idx < num_params:
+                shard_sig = _gen_shard_strategy(mesh, 0)
+                replica_sig = _gen_replicate_strategy(mesh)
```
I assume the above two lines mean the param can either be distributed using FSDP (`shard_sig`) and DDP (`replica_sig`)?

For `shard_sig`, do we need to find the best dim to shard in follow up PRs?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                # the corresponding parameters
+                # NOTE: optimizer states can be None sometimes,
+                # creating a NON_TENSOR stratgy for those cases.
+                if "val" not in node.meta:
```
IIUC, this is before calling into `partitioner()`. If so, does `make_fx` sets this field for us automatically?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                # strategy for those cases.
+                if "val" not in node.meta:
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.NON_TENSOR, []
```
nit: this can be consolidated with the `NON_TENSOR` tagging of optimizers?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+
+    # next we forward propagate to mark all the sharding
+    for node in train_step_graph.graph.nodes:
+        if node.op == "placeholder":
```
nit: we have `OP.PLACEHOLDER` in `torch.distributed._spmd.graph_utils`

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+            # param/state + grad (param/state/grad be the same spec) -> out: param/state
+            # param + state -> out: param
+
+            if node.target in non_compute_ops:
```
at this point, i assume `tag_grad` nodes should all have been removed. If that's the case, then both usage of `non_compute_ops` don't expect to see `tag_grad` in there?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                    len(input_nodes) == 1
+                ), f"non-compute op only support one input now, found node: {node} with length of inputs: {len(node.args)}"
+                arg_node_type = dp_strategy_map[input_nodes[0]].node_type
+                if dp_strategy_map[input_nodes[0]].reduction_over_batch:
```
nit: looks like we can directly do the following and save line 359?

```python
input_full_reduction = dp_strategy_map[input_nodes[0]].reduction_over_batch
```

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.ACT, [shard_sig]
+                    )
+                else:
```
missing `arg_node_type  = NodeType.STATE`?

wanchaol replied:
I am not sure how to handle non compute ops to optimizer states yet, maybe we can see what make sense if we hit such case.

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+            # for computatation nodes, we need to check all the inputs
+            input_args = node.all_input_nodes
+            input_node_types = pytree.tree_map(
+                lambda x: dp_strategy_map[x].node_type, input_args
```
nit: `all_input_nodes` already returns a list of nodes, we can save flatten and unflatten from `tree_map` and directly do `input_node_types = [dp_strategy_map[a].node_type for a in input_args]`?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                # found a param_grad node that already have output pre-filled spec
+                # fill in the expected input specs for the pre-filled strategy
+                node_type = dp_strategy_map[node].node_type
+                assert node_type == NodeType.GRAD
```
I assume this is thanks to `tag_grad` and line 288's backward prop

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+        shape = node.meta["val"].shape
+
+        # for reduction op that reduces over the sharded batch dim
+        # we don't generate partial, but rather, we generate shard
```
This should work, but seems to be slightly abusing DTensor's placement definitation? E.g., if I do `sum()` and get a scalar tensor, the tensor itself actually does not have a batch dimension to shard?

wanchaol replied:
Oh yeah you are right. that actually need to be specifically handled, I used a `from_local` flag in the partitioner to indicate that this node don't really have dim and so it can skips the sharding.

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                        elif node_type == NodeType.PARAM:
+                            # param must be replicated
+                            input_specs.append(
+                                DTensorSpec(mesh=mesh, placements=[Replicate()])
```
I assume this is how we trigger FSDP's allgather?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                            )
+                        else:
+                            raise RuntimeError(
+                                f"Expecting node with parameter and activation, but found {input_node_types}! "
```
(can be done in follow up PRs): ops like `nll_loss` does take two activation tensors as input?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+            continue
+
+        if node.op == "placeholder":
+            # print(f">>>> processing placeholder: {node}")
```
nit: debugging vestige

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                    replica_strategy = _gen_replicate_strategy(mesh)
+                    shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                    output_node_type = NodeType.PARAM
+                    for arg_node_type in input_node_types:
+                        if arg_node_type != NodeType.GRAD:
+                            output_node_type = arg_node_type
+                            break
+                    assert (
+                        output_node_type == NodeType.STATE
+                        or output_node_type == NodeType.PARAM
+                    ), "Expecting output node type to be either state or param! "
```
nit:

```python
non_grad_types = [t for t in input_node_types if t != NodeTypeGrad]
assert len(non_grad_types) == 1 and non_grad_types[0] in [NodeType.STATE, NodeType.PARAM], "..."
```

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+                    replica_strategy = _gen_replicate_strategy(mesh)
+                    shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                    output_node_type = (
+                        NodeType.PARAM
```
is this guaranteed to be true, i.e., output is guaranteed to updated param as long as param is in the args?

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+            if len(node_strategies) == 1:
+                node.meta["sharding"] = node_strategies[0]
+            elif len(node_strategies) == 2:
+                if parallel_mode == DataParallelStyle.REPLICATE:
+                    # set to replicate for replicate style
+                    node.meta["sharding"] = node_strategies[0]
+                elif parallel_mode == DataParallelStyle.FULLY_SHARD:
+                    # set to shard for fully shard style
+                    node.meta["sharding"] = node_strategies[1]
```
IIUC, this means global `parallel_mode` dictates the style. I assume we plan to do smarter things later, otherwise, we don't need to generate `node_strategies` list?

If yes, let's add a TODO/FIXME

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,763 @@
+
+def mark_data_parallel_shardings(
+    train_step_graph: GraphModule,
+    num_parameters: int,
+    num_states: int,
```
Yeah true, we can check the node types instead, I'll make a refactor as a follow up PR

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

mrshenli reviewed (commented):
will come back for the rest tmr :)

mrshenli reviewed (approved):
LGTM! Lint error is real, please address before landing. Thanks!

wanchaol commented on torch/distributed/_spmd/api.py:
```diff
@@ -429,22 +438,47 @@ def stateless_func(func, params, buffers, named_states, args, kwargs):
+            fake_arg = fake_mode.from_tensor(arg)
+            arg_dims = [1] * arg.ndim
+            # we assume the first dim is batch dim in data parallel
+            arg_dims[0] *= dist.get_world_size()
```
Yes, the aten.expand would use the global batch size for grad calculation (and also the whole graph). But when we goes into really partition the graph (in partitioner), the output of aten.expand is sharded on batch, so we would adjust the shape to local batch size at that time.

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45174400).

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+
+_spmd_lib_impl = torch.library.Library("_spmd", "IMPL")
+for dispatch_key in ("CPU", "CUDA", "Meta"):
+    _spmd_lib_impl.impl("tag_grad", lambda x: x, dispatch_key)
```
Instead of registering each dispatch key separately, you can register this to CompositeExplicitAutograd

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                    is similar to the behavior of FullyShardedDataParallel, the
+                    difference is that FullyShardedDataParallel (ZERO-3), which
+                    shards the model using FlatParameter based sharding,
+                    while this style shard each parameter into DTensor.
```
nit: shards

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+    ACT = 1
+    GRAD = 2
+    STATE = 3
+    NON_TENSOR = 4  # NON_TENSOR is to tag non tensor node (i.e. graph output)
```
cc @bdhirsh this is pretty similar to the input/output tagging we've been discussing recently

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+    def __init__(
+        self,
+        node_type: NodeType,
+        startegy_list: List[PlacementStrategy],
```
nit: strategy_list

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+    finally:
+        # remove those hooks after tracing
+        for h in tagging_hooks:
+            h.remove()
```
No change needed now, but this doesn't feel like it should be necessary, it should be "obvious" at trace time which nodes are grads versus not.

ezyang replied:
OK, so I think I initially misunderstood the function of this. You are only adding these to the final gradient for every parameter, rather than every grad value flowing through the graph. And the reason for this is that you're tracing a full train-step graph, which means that the gradients aren't actually outputs of your graph, they immediately get fed to the optimizer.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+            h.remove()
+
+
+class BatchDimAnalyzer(object):
```
nit: You don't need to explicitly inherit from object

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+
+    def get_batch_dim_shard_spec(
+        self, node: fx.Node, mesh: DeviceMesh, input_full_reduction: bool = False
+    ) -> Tuple[DTensorSpec, bool]:
```
It feels like this can be factored into two functions: one function that computes the batch dim, and then another which computes the sharding spec from the batch dim.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+            self.set_batch_dim(node, self.get_batch_dim(operand))
+        else:
+            for i, dim_size in enumerate(shape):
+                if dim_size == self.batch_dim_size:
```
This looks very questionable. At the very least, you should not be willing to mark MULTIPLE dims as the batch dim...

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+def build_data_parallel_strategies(
+    train_step_graph: GraphModule,
+    num_params: int,
+    num_states: int,
```
What are the meaning of these two parameters?

wanchaol replied:
These two parameters are `number of parameters` and `number of optimizer states`. It's used to determine the boundary of placeholders so that we could mark parameter placeholders and `PARAM` and optimizer states placeholders as `STATE`

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+    num_params: int,
+    num_states: int,
+    mesh: DeviceMesh,
+    batch_dim: int = 0,
```
What exactly does this mean? Is it assumed that all inputs are batched at this dim?

wanchaol replied:
Yeah it's a implicit assumption that the input is a mini-batch that is sharded on batch dimension 0 (i.e. a full batch input is sharded on dim 0 to get mini-batch input on each rank), it's being used to initialize the `BatchDimAnalzyer`. This implicit assumption might be too strong, so I pass this as a parameter and plan to expose this to the `DataParallelMode` directly so that it could support cases like batch_dim is not 0-dim.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+        aten.view.default,
+        aten.reshape.default,
+        aten.detach.default,
+        aten.clone.default,
```
What makes something non-compute? I see both view and non-view operations in this list. Is this anything that's bandwidth bound?

ezyang replied:
I guess, most of these make sense except `ones_like`, that's surprising.

wanchaol replied:
Mostly a non-compute ops does not really do computation and only do some simple transformations (view, reshape, etc). yeah let me re-think `ones_like` and how to handle it

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+
+            if num_param_grad == num_params:
+                # early break if we have already processed all param_grads
+                break
```
Debug assert that there are no more `tag_grad` in graph at this point?

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                # early break if we have already processed all param_grads
+                break
+
+    train_step_graph.recompile()
```
If you're not going to use the Python code (e.g., by `__call__` the graph module), recompiling here is pointless and wasting time. You only need to recompile at the very end.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                # 1. optimizer states can be None sometimes, i.e. SGD with
+                #    no momentum, optimizer states populate `momentum` state
+                #    as None, the full graph we get from `compile` would have
+                #    None as the placeholder value
```
Hmm, but then why do we have a node for it at all? If something is None I would expect it to just not show up in the graph.

wanchaol replied:
That's a good point actually. I think it might be due to how we lift the optimizer states, currently we lift all optimizer states, maybe we should not lift the optimizer states if they are just None. I'll create a follow up PR to address it.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+
+            elif placeholder_idx < num_params:
+                # during compilation there's an assumption that the first num_params
+                # placeholders should be parameters
```
@bdhirsh This is a good example of the utility of input specs, otherwise people have to manually work out the calling convention.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                shard_strategy = _gen_shard_strategy(mesh, 0)
+                replica_strategy = _gen_replicate_strategy(mesh)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.PARAM, [replica_strategy, shard_strategy]
```
noob question: what exactly does this mean? Specifically, the list of strategies passed to DataParallelStrategy. Is this saying that the 0th dim is replicated and the 1st dim is sharded? (That doesn't seem right). Or is it saying that we should try to replicate it first, and if that doesn't work, shard it? (But that also doesn't seem right?)

wanchaol replied:
I should clarify more in the `DataParallelStrategy` comment, will add it. Basically a `PlacementStrategy` means one way to distribute the tensor, a StrategyList(DataParallelStrategy) that have a `List[PlacementStrategy]` means there could be multiple ways to distribute the tensor. 

In this data parallel case, for model parameters, there're two possible ways to distribute the parameters: 1. replicate the parameter over a set of devices (DDP like behavior) 2. shard the parameter on its tensor dimension 0 over a set of devices (FSDP like behavior). 

So here we create a `DataParallelStrategy` that have two possible ways of sharding this parameter during `BuildDataParallelStrategy`. Then later in `mark_data_parallel_sharding`, according to whether the DataParallelStyle is replicate or fully_shard, we assign only one `PlacementStrategy` (replicate use the `replica_strategy`, fully_shard use the `shard_strategy`) to the node and use this strategy to do the real sharding.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+            elif placeholder_idx < activation_idx:
+                # optimizer states follow the same strategy as
+                # the corresponding parameters
+                placemenet_strategies: List[PlacementStrategy] = []
```
nit: placement_strategies

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+            else:
+                activation_batch_dim_size = node.meta["val"].shape[batch_dim]
+                # find the first activation node and use its batch dim size
+                if batch_dim_analzer.batch_dim_size == -1:
```
nit: batch_dim_analyzer

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                shard_strategy = _gen_shard_strategy(mesh, batch_dim)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.ACT, [shard_strategy]
+                )
```
Isn't the graph a full forward-backward-optstep graph? How come activations are inputs?

wanchaol replied:
Yes it's a fwd+bwd+optim graph. I didn't distinguish inputs and activations and treated them to have the same `NodeType` as I would like to keep the data parallel rule be relatively simple.

ezyang commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                    NodeType.ACT, [shard_strategy]
+                )
+            placeholder_idx += 1
+        elif node.op == "call_function":
```
Meta comment: the propagation logic here is very long. It feels like there should be some way to make it shorter

ezyang replied:
The length makes it difficult to tell which rules are "special cased" for FSDP, versus which rules would have worked in a more generic DTensor propagation environment

wanchaol replied:
Thanks let me try and see how to split this in a cleaner way!

wconstab commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+            operand = node.all_input_nodes[0]
+            if operand in self.batch_dim_map:
+                operand_batch_dim = self.get_batch_dim(operand)
+                if operand_batch_dim == 0:
```
is this missing handling for a reduction op that doesn't reduce over dim=0?

wconstab commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                if arg_node_type == NodeType.PARAM:
+                    replica_strategy = _gen_replicate_strategy(mesh)
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.PARAM, [replica_strategy]
```
a little confused, wouldn't Param nodes be sharded, and that implies that a non-compute op (aka a view) of the param would also be sharded?

wanchaol replied:
For `placeholders` (which are params), it could either be `replicated/sharded`, and it corresponds to DDP/FSDP style. But for `call_function` nodes (whether it's a view/non-compute or compute node), we always treat `PARAM` nodes as `Replicate`, this is because during computation it would always need the parameter be fully replicated (where it would trigger allgather operation to turn sharded param to replicated during `call_function` execution)

wanchaol commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -0,0 +1,795 @@
+                if node.meta["val"].ndim > 0:
+                    # if the state is a scalar, we don't want to shard it
+                    # i.e. param state step is usually a scalar tensor
+                    shard_strategy = _gen_shard_strategy(mesh, 0)
```
Yeah agreed, this is actually the `DEFAULT` DataParallelStyle would do, it would need to actually look into the `param` shape to determine whether the param can be sharded evenly or not, or the param is too small to shard, I will do it as a follow up PR. I think the `build_data_parallel_strategies` would still generate `Replicate()` and `Shard(tensor_dim=0)` as a strategy, to determine which one to really use we can probably do it in `mark_data_parallel_sharding`

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

Differential Revision: [D45174400](https://our.internmc.facebook.com/intern/diff/D45174400)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45174400).

wanchaol committed:
Update on "[spmd] Introduce Compile Mode FSDP with DTensor"


This PR introduces compile mode Data Parallel (FSDP/DDP) using DTensor sharding.

Along with the algorithm, it also introduces a new DataParallelMode so that `compile` API can take it
and apply data parallel. This PR trys to preserve the DTensorExpand
approach first to avoid BC, we shall discuss steps to remove
DTensorExpand.

The data parallel mode uses heuristics to determine node types in the
graphs and assign the corresponding sharding. The detailed algorithm
described in the design doc.

The benefits of this approach:
- Model parameters and optimizer states are all DTensors after  `spmd.compile`, which is necessary for FSDP, and also makes it super easier for checkpointing
- As model parameter/optim states are sharding in a per-parameter approach, it would be able to compose with sophisticated second order optimizer (i.e. Shampoo) in a easier way.
- We leverage the model parameter/grads information to derive data parallel pattern. In this way we don't need to worry about DTensor op coverage anymore! As data parallel is just a special case of DTensor operation.
- Use dtensor_expand might work for DDP but aren't going to work for FSDP as dtensor might choose to allgather activation, which might violate native fsdp algorithm.
- The approach is general enough to support both DDP/FSDP and a mixed mode

Follow ups:
- Add the "default" data parallel mode which supports mixing of
replicate/fully shard
- Test more e2e models with more different types of optimizers, etc
- migrate the existing stack from the DTensorExpand mode
- build optimizations on top of this prototype

Differential Revision: [D45174400](https://our.internmc.facebook.com/intern/diff/D45174400)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45174400).

awgu commented on torch/distributed/_spmd/parallel_mode.py:
```diff
@@ -50,6 +54,75 @@ def transform_and_compile(self, gm: GraphModule) -> GraphModule:
+
+    def __init__(
+        self,
+        parallel_style: str = "replicate",
```
Why is `parallel_style` a `str` instead of enum?

wanchaol replied:
`DataParallelStyle` is actually a Enum defined inside `data_parallel.py`, and we used it here inside the constructor too. The reason I choose to use `str` in the constructor is that I don't quite like a `Enum` to be a public facing API as that requires an additional import statement imposed to users, especially when it's just an option with 2-3 choices. 

We should probably study the UX later to decide which is better before the release, this is right now just my preference.

awgu commented on test/distributed/_spmd/test_data_parallel.py:
```diff
@@ -0,0 +1,163 @@
+                p.grad *= self.world_size
+        ddp_opt.step()
+
+        # compile it with replicate and run step once
```
nit: The comment is out of sync with code since `data_parallel_mode` may not be `"replicate"`.

wanchaol replied:
thanks, let me create a follow up PR with the test refactor together to avoid triggering CI again.

awgu commented on test/distributed/_spmd/test_data_parallel.py:
```diff
@@ -0,0 +1,163 @@
+        torch.sum(ddp_mod(ddp_inp[0]) - ddp_inp[1]).backward()
+        with torch.no_grad():
+            for p in ddp_mod.parameters():
+                p.grad *= self.world_size
```
Curious: Is there a reason why compile-mode data parallel does not average gradients by world size?

wanchaol replied:
I was following the other tests in `test_tracing.py` where it does not average gradients for the DTensorExpandMode, so honestly idk why we didn't do it, maybe @mrshenli have a better idea there. If there's no specific reason, I think we probably should average the gradients by default, I'll do it in a follow up PR.

awgu commented on test/distributed/_spmd/test_data_parallel.py:
```diff
@@ -0,0 +1,163 @@
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_fully_shard_sgd(self):
```
nit: Maybe `test_fully_shard_sgd` and `test_fully_shard_sgd_momentum` can be refactored into a helper that takes in the optimizer kwargs to avoid so much duplication.

wanchaol replied:
Yeah I need to think a bit how to refactor the tests to avoid duplication, I'll create a follow up PR to clean it up.

wanchaol commented:
Thanks for all the reviews! Addressed most of comments, and there're still a bunch I want to address. Given that this PR is becoming quite large, I'll land this first and address the remaining comments in separate PRs :)

wanchaol commented:
@pytorchbot merge
